src-c/native/stubs.c

```diff
@@ -24,10 +27,16 @@
                                                                              \
   CAMLprim value                                                             \
   caml_digestif_ ## name ## _ba_update (value ctx, value src, value off, value len) { \
-    digestif_ ## name ## _update (                                           \
-      (struct name ## _ctx *) String_val (ctx),                              \
-      _ba_uint8_off (src, off), Int_val (len));                              \
-    return Val_unit;                                                         \
+    CAMLparam4 (ctx, src, off, len);                                         \
+    uint8_t *off_ = Caml_ba_data_val(src) + Int_val (off);                   \
+    int len_ = Int_val (len);                                                \
+    struct name ## _ctx ctx_;                                                \
+    memcpy(&ctx_, String_val(ctx), sizeof(struct name ## _ctx));             \
+    caml_release_runtime_system();                                           \
+    digestif_ ## name ## _update (&ctx_, off_, len_);                        \
+    caml_acquire_runtime_system();                                           \
+    memcpy(String_val(ctx), &ctx_, sizeof(struct name ## _ctx));             \
```

Comment:
Padding data will be copied as is - i.e. the value as was when it was constructed. So there is no information leak here. 

IMO a more 'correct' way would be to define `ctx` as a custom type, and add a function to copy it (to implement dup) - But I don't know if `ctx` is supposed to be interchangeable with the ocaml implementation. 

I could change the implementation to copy the struct using the assignment operator:
```c 
struct name ## _ctx ctx_ = *( (struct name ## _ctx*) Bytes_val(ctx));`
...
*( (struct name ## _ctx*) Bytes_val(ctx)) = ctx_;
```
Which in theory would only copy the needed fields (although I'm sure what the standard mandates in this case).

Reply:
This [StackOverflow answer](https://stackoverflow.com/a/10954584) seems to suggest that the only way to be sure not to leave padding bytes is to assign field-by-field, so a function or macro for each `ctx` types sounds like it would be the "correct" way to go, as you suggested.
I don't know if this is necessary, or even if we're using C99 and this applies to us (I never really learned C or C build systems, so I have no idea if this is the case).

If we all agree on the invariant that we clear the ctx padding on initialization, and/or never expose the padding to OCaml users, I think we're fine. Maybe this would be  good place to add a comment documenting that invariant?

@andersfugmann sorry to start this discussion on your PR; it's not a criticism of your code, I just never looked into the C implementation before.

Reply:
It seems that at any initialization functions for each hash implementations, we do a `memset` to set the context to zero. It's like:

```
memset( ctx, 0, sizeof( struct hash_ctx ) );
```

So padding bytes __are not__ set to 0 if I understand correctly post pointed by @cfcs. On top of it, `stubs.c` does a C (unsafe) cast between a `String_val` and the proper `ctx`. I agree about the proposition to make, for each impl., a `dup` function properly cast from a `String_val` to a `ctx`. I think, if @cfcs and @andersfugmann are happy about that, I can implement it in an other PR.

Reply:
@dinosaure the padding bytes as far as I can tell *are* set to zero on initialization by the `memset`.

With the `memcpy` here we copy it back to an ocaml `bytes` values, but since the padding is always zero (we don't copy other stuff into the padding bytes), this is OK in practice.
It just means that we have a hidden invariant that we cannot set or initialize these contexts differently.
If we provided a `ctx_of_string` function somewhere (like for resuming a long-running hashing operation from disk or whatever) this would be more of an issue since we then either need to validate that the padding is zero'ed (*eww!*) or end up with a type where a by-value comparison of two "actually equal" `ctx`s will fail because of the padding.

Having dedicated functions for copying the fields seems to be the only way to deal with this safely in C, which is a bit of a bummer. I'm not sure it's worth adding the complexity.